cmd/autoretrieve/main.go

```diff
@@ -0,0 +1,772 @@
+package main
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/application-research/filclient"
+	"github.com/application-research/filclient/keystore"
+	"github.com/application-research/filclient/retrievehelper"
+	"github.com/dustin/go-humanize"
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-fil-markets/retrievalmarket"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/wallet"
+	lcli "github.com/filecoin-project/lotus/cli"
+	bsmsg "github.com/ipfs/go-bitswap/message"
+	bitswap_message_pb "github.com/ipfs/go-bitswap/message/pb"
+	bsnet "github.com/ipfs/go-bitswap/network"
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	flatfs "github.com/ipfs/go-ds-flatfs"
+	leveldb "github.com/ipfs/go-ds-leveldb"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	"github.com/libp2p/go-libp2p-kad-dht/fullrt"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/xerrors"
+)
+
+var logger = logging.Logger("estuary-ar")
+
+func main() {
+	logging.SetLogLevel("estuary-ar", "DEBUG")
+
+	app := cli.NewApp()
+
+	app.Flags = []cli.Flag{
+		&cli.StringFlag{
+			Name:    "datadir",
+			Value:   "./estuary-ar",
+			EnvVars: []string{"ESTUARY_AR_DATADIR"},
+		},
+		&cli.IntFlag{
+			Name:  "timeout",
+			Value: 60,
+			Usage: "Time in seconds to wait on a hanging retrieval before moving on",
+		},
+		&cli.StringFlag{
+			Name:  "endpoint",
+			Value: "https://api.estuary.tech/retrieval-candidates",
+		},
+	}
+
+	app.Commands = []*cli.Command{
+		{
+			Name: "test-blacklist",
```

Comment:
you could just call this 'check-blacklist' and leave it as a permanent part of the codebase, no need for it to be a 'test' command. I generally like to leave all sorts of utility commands around